smart-contract/src/contract.rs

```diff
@@ -404,114 +405,132 @@ fn try_redeem(
-            let wamt = if c.amount.u128() >= remaining {
-                remaining
-            } else {
-                c.amount.u128()
-            };
-            // Withdraw max amount from the requested marker.
-            res = res
-                .add_message(withdraw_coins(
-                    &reserve_denom,
-                    wamt,
-                    &reserve_denom,
-                    info.sender.clone(),
-                )?)
-                .add_attribute("withdraw_amount", Uint128::new(wamt))
-                .add_attribute("withdraw_denom", &reserve_denom);
-            // Decrement the pending amount due
-            remaining -= wamt;
-        }
+    // Validate params
+    if amount.is_zero() {
+        return Err(contract_err("invalid redeem and burn amount"));
     }
 
-    // Try and withdraw from other reserve markers if total due hasn't been withdrawn.
-    if remaining > 0 {
-        // Query markers with non-zero balances excluding the requested denom
-        let bres = get_balances(deps.as_ref())?;
-        let balances: Vec<Balance> = bres
-            .balances
-            .iter()
-            .filter(|b| b.denom != reserve_denom && !b.amount.is_zero())
-            .cloned()
-            .collect();
-        // Process withdrawals
-        for balance in balances.iter() {
-            if remaining > 0 {
-                // Determine the max amount that can be withdrawn.
-                let wamt = if balance.amount.u128() >= remaining {
-                    remaining
-                } else {
-                    balance.amount.u128()
-                };
-                // Withdraw available amount from the requested marker.
-                res = res
-                    .add_message(withdraw_coins(
-                        &balance.denom,
-                        wamt,
-                        &balance.denom,
-                        info.sender.clone(),
-                    )?)
-                    .add_attribute("withdraw_amount", Uint128::new(wamt))
-                    .add_attribute("withdraw_denom", &balance.denom);
-                // Decrement the pending amount due
-                remaining -= wamt;
-            }
-        }
+    // Read state
+    let state = config_read(deps.storage).load()?;
+
+    // Ensure message sender is a member.
+    let mut member = members_read(deps.storage).load(info.sender.as_bytes())?;
+
+    // Ensure member holds at least the requested amount of dcc tokens.
+    let balance = deps.querier.query_balance(&member.id, &state.dcc_denom)?;
+    if balance.amount < amount {
+        return Err(contract_err("insufficient dcc token balance"));
     }
 
-    // If there is still a due amount, we can't meet the request so fail the tx
-    if remaining > 0 {
-        return Err(contract_err("insufficient reserve tokens in escrow"));
+    // Use the member's reserve denom.
+    let reserve_denom = member.denom.clone();
```

Comment:
It appears this extra allocation isn't needed. You should be able to use `&member.denom` directly below.

Reply:
As is in this case it's just an unnecessary extra string allocation. In grant scheme of things it's very minor. Here's an example that compiles if you did want to remove it.

Reply:
yeah i switched to just referencing member.denom directly